integrate telemetry

annthurium opened the pull request:
this pull request integrates the new `telemetry` package for sending metrics to GitHub's internal analytics pipeline, rather than Google Analytics. For more context on this project: http://blog.atom.io/2018/06/20/atom-metrics.html

More specifically:
- adds the new `telemetry` lib as a dependency.
- instantiates a `StatsStore` to hold data which will eventually be sent to GitHub's internal analytics piepline.
- adds some methods to the metrics reporter service that expose `telemetry` apis directly so that other Atom packages can consume them.
- modifies the existing methods of the metrics reporter service so that when they are called, data is added to `telemetry`.  The plan is, for a while we'll be in a state where we are sending data to both Google Analytics and the internal pipeline.  That way, we can cross check both sources.  When we are confident that the internal pipeline data is solid, we will deprecate sending data to Google Analytics.
- adds unit tests for new functionality
- increases unit test coverage for existing code a tiny bit.
  - tests opt in / opt out events
  - tests timing events
- clarifies language in some of the tests.

Manual testing plan:

- [x] Verify that data is still being sent to Google Analytics from Atom using the network tab of the console when new files are opened.
- [x] Manually call `incrementCounter` and `addCustomEvent` from the console. Manually call `store.reportStats()` and verify that the stats are successfully sent to `central`. 
- [x] verify that Atom tries to send stats when the reporting loop interval has passed

sguthals commented:
Could you give guidance on what you're looking back in terms of feedback? 

My assumption is, at minimum you're looking for:
@jasonrudolph to give an overview of the code and provide feedback
@sguthals to be aware of what is being built and the scope
Are these assumptions correct?

For context, in the future, a review ask might also include "can you dogfood this" or "can you test this thoroughly"

jasonrudolph commented on spec/metrics-spec.js:
```diff
@@ -142,13 +173,14 @@ describe('Metrics', async () => {
 
         const {mainModule} = atom.packages.getLoadedPackage('metrics')
         mainModule.shouldIncludePanesAndCommands = true
+        Reporter.addCustomEvent.reset()
```
Given that we're [calling this in the `afterEach` function](https://github.com/atom/metrics/blob/a9afa070ff8b1d82815012f9a1da0c4c369a86b0/spec/metrics-spec.js#L34) for all tests in this file, can we remove this line, or do we need it?

annthurium replied:
💥

annthurium replied:
Removed the reset from all tests in this file, as it's unnecessary, and left it here.

jasonrudolph commented on spec/metrics-spec.js:
```diff
@@ -320,29 +362,42 @@ describe('Metrics', async () => {
       })
 
       it('will not report pane items', async () => {
+        Reporter.addCustomEvent.reset()
```
Given that we're [calling this in the `afterEach` function](https://github.com/atom/metrics/blob/a9afa070ff8b1d82815012f9a1da0c4c369a86b0/spec/metrics-spec.js#L34) for all tests in this file, can we remove this line, or do we need it?

annthurium replied:
ah, these were some debug code trying to get to the bottom of my comment re file open events.  Will remove them.

jasonrudolph commented on spec/metrics-spec.js:
```diff
@@ -232,28 +263,28 @@ describe('Metrics', async () => {
     describe('when there are paths in the exception', () => {
       it('strips unix paths surrounded in quotes', () => {
-        let message = "Error: ENOENT, unlink '/Users/someguy/path/file.js'"
+        let message = "Error: ENOENT, unlink '/Users/someuser/path/file.js'"
```
:zap:

jasonrudolph commented on lib/reporter.js:
```diff
@@ -37,6 +38,13 @@ const getOsArch = function () {
+  static addCustomEvent (event, eventName) {
+    store.addCustomEvent(event, eventName)
+  }
   static sendEvent (category, action, label, value) {
```
I'm guessing that you'd want anyone writing new code to use `addCustomEvent` instead of `sendEvent`. Is that correct? If so, I'd recommend marking `sendEvent` as deprecated [[example](https://github.com/atom/atom/blob/bee32d58d9b73d7eff70b83086c9a18033ea46aa/src/text-editor-registry.js#L179)]. What do you think?

daviwil replied:
Also, could/should it be called `sendCustomEvent`?  To me, `add` sounds like we're registering an event type rather than sending some information

annthurium replied:
ah, good call.  Will change this too.

annthurium commented:
@sguthals you asked to be added to code reviews as a non blocking reviewer awhile back.  I didn't have specific feedback I wanted from you with regards to this pull request, as I know I can count on @jasonrudolph for a thorough code review.  😄

sguthals commented:
Awesome! That is what I assumed 😄 

In general, I'd like to get in the habit of including in the PR description what you're looking for in terms of feedback. At some point it will be nice to provide feedback on things like usability) and if we get in the habit of being specific about the type of feedback we're looking for I think both pr author and reviewer will be happier 👍

daviwil commented on lib/reporter.js:
```diff
@@ -96,6 +109,7 @@ class Reporter {
       ev: this.commandCount[commandName]
     }
 
+    this.addCustomEvent(params, 'command')
```
I agree, having the event name first makes more sense to me and seems to be common practice in API design.  Typically my principle for argument ordering is "most important/relevant parameter first" and I think in this case the event name is the most important parameter.

annthurium replied:
sure thing!  can make the event name first.

daviwil commented on lib/store.js:
```diff
@@ -0,0 +1,7 @@
+const telemetry = require('telemetry-github')
+
+const store = new telemetry.StatsStore('atom', atom.getVersion(), atom.inDevMode(), () => '')
```
I'd also make that parameter optional and do the "right thing" when the user doesn't pass anything.

annthurium replied:
great.  fixed.  thank you!

annthurium commented on spec/metrics-spec.js:
```diff
@@ -20,7 +29,29 @@ describe('Metrics', async () => {
+    spyOn(store, 'setOptOut')
+    spyOn(Reporter, 'sendEvent')
+    await atom.config.set('core.telemetryConsent', 'limited')
+
```
ah, good catch.  fixed.

annthurium commented on spec/metrics-spec.js:
```diff
@@ -168,36 +208,27 @@ describe('Metrics', async () => {
-
-        atom.commands.dispatch(workspaceElement, 'editor:move-to-end-of-line', null)
-        expect(Reporter.request).not.toHaveBeenCalled()
+        assertNotCalledHelper('core:move-up')
```
👍

annthurium commented on lib/store.js:
```diff
@@ -0,0 +1,7 @@
+const telemetry = require('telemetry-github')
+
+const store = new telemetry.StatsStore('atom', atom.getVersion(), atom.inDevMode(), () => '')
+const hasOptedOut = atom.config.get('core.telemetryConsent') !== 'limited'
```
I could call it `notOptedIn` because that's really what `telemetry` cares about.  Will update.

annthurium commented on lib/reporter.js:
```diff
@@ -71,15 +81,18 @@ class Reporter {
+    const eventName = 'appview'
     const params = {
-      t: 'appview',
+      t: eventName,
```
cool.  done.

jasonrudolph commented on lib/reporter.js:
```diff
@@ -58,6 +67,7 @@ class Reporter {
     }
 
     this.send(params)
+    // todo (tt, 6/2018): send timing to telemetry.
```
@annthurium: As a general guideline, my personal preference is to track TODOs in issues, as opposed to code comments. 😺

annthurium replied:
cool.  I sometimes add TODOs if I think other folks reading/modifying this code would want to be aware of imminent changes.  Fine with removing this though.

annthurium commented on package-lock.json:
```diff
@@ -0,0 +1,1912 @@
+      "version": "0.0.33",
+      "resolved": "https://registry.npmjs.org/tmp/-/tmp-0.0.33.tgz",
+      "integrity": "sha512-jRCJlojKnZ3addtTOjdIqoRuPEKBvNXcGYqzO6zWZX8KfKEpnGY5jfggJQ3EjKuu8D4bJRr0y+cYJFmYbImXGw==",
+      "dev"
```
oh, thanks! I had no idea I was not supposed to check in package.lock files for Atom packages.

Arcanemagus replied:
Since `apm` currently ignores them they aren't giving us any benefit right now, and simply add a bunch of extra maintenance work at the moment.

annthurium commented:
@jasonrudolph : I believe I've addressed all the comments.  Thank you for your feedback!  Would you be willing to take a second pass through this?  Given that this code is important and I'm still new, I'd like an explicit 👍 before I merge.

As far as what kind of feedback I'm looking for:
- does my naming make sense?
- are there any style nits that would enhance readability / consistency?
- is there anything I should be unit testing but I'm not?
- is there anything missing from the manual testing plan?

annthurium commented on package.json:
```diff
@@ -18,7 +18,8 @@
   "dependencies": {
     "fs-plus": "^3.0.0",
     "grim": "^2.0.1",
```
@jasonrudolph : I'm assuming that I don't need to increment the version of `metrics-reporter` in `providedServices` here since I'm just adding new methods and the old ones are still supported.  Is that the case?

jasonrudolph replied:
Since this is an additive change, I think it makes sense to increment the version from `1.0.0` to `1.1.0`.

Since the new methods didn't exist in version `1.0.0` of the service, and since you want a way for packages to consume the new methods, those packages need to be able to declare a dependency on a version of the service that includes the new methods. By incrementing the version to `1.1.0`, existing consumers (e.g., the [`welcome` package](https://github.com/atom/welcome/blob/7e504a6895f6123942ed6dbb219dd564604e5285/package.json#L19)) should continue to work, and packages that want to use the new methods can declare a minimum dependency of `1.1.0`.

Does that sound reasonable?

annthurium replied:
got it, thanks!  Have incremented the `provideReporter` version here.

annthurium commented:
Thanks for the second pass, @jasonrudolph!

To answer your questions about the test plan:

> All of the stated steps call for using Atom in dev mode. Is there something noteworthy about dev mode as it relates to testing this functionality? Or would you expect these verification steps to also succeed when running Atom in normal (non-dev) mode?
Correct - these steps would also succeed in non-dev mode.  I suppose specifying that I tested in dev mode was rather pedantic of me. 😛 I am happy to change it if you think it could be confusing.

>As far as I can tell, there's no production code that currently uses incrementCounter. Will you manually invoke incrementCounter and verify that it has the desired effect?
Correct, I will manually invoke incrementCounter.

>Am I correct in thinking that sendException will continue to send info only to Google Analytics, and will intentionally not send data to Telemetry?
That is also correct - `telemetry` doesn't handle exceptions.  I wanted to get the metrics working first and then worry about exceptions.

annthurium commented:
yay, it works, it works!!  

<img width="407" alt="screen shot 2018-06-29 at 3 31 52 pm" src="https://user-images.githubusercontent.com/3781742/42117495-a6d094aa-7bb1-11e8-9258-b04f4b6ded72.png">

thanks again @jasonrudolph for all your help with this!

![bitmoji](https://render.bitstrips.com/v2/cpanel/15f269e9-79ff-42f5-b6f8-df81e1cb3847-f14a1e1d-64bc-41f2-8006-01b6c63adbd8-v1.png?transparent=1&palette=1&width=246)
